presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTokenSplitManager.java

```
@@ -13,129 +13,132 @@
-        List<Future<List<TokenSplit>>> splitFutures = new ArrayList<>();
-        for (TokenRange range : masterRangeNodes) {
-            // for each range, pick a live owner and ask it to compute bite-sized splits
-            splitFutures.add(executor.submit(new SplitCallable<>(range, keyspace, columnFamily, splitSize, cassandraThriftClient, partitioner)));
-        }
+        this.session = requireNonNull(session, "session is null");
+        this.splitSize = splitSize;
+    }
 
-        // wait until we have all the results back
-        for (Future<List<TokenSplit>> futureInputSplits : splitFutures) {
-            try {
-                splits.addAll(futureInputSplits.get());
+    public List<TokenSplit> getSplits(String keyspace, String table)
+            throws IOException
+    {
+        Set<TokenRange> tokenRanges = getTokenRanges();
+        List<SizeEstimate> estimates = getSizeEstimates(keyspace, table);
+        Map<TokenRange, SizeEstimate> estimateMap = mapSizeEstimates(estimates);
+
+        ImmutableList.Builder<TokenSplit> splits = ImmutableList.builder();
+
+        for (TokenRange tokenRange : tokenRanges) {
+            List<String> endpoints = unmodifiableList(getReplicas(keyspace, tokenRange)
+                    .stream()
+                    .map(Host::toString)
+                    .collect(toList()));
+
+            SizeEstimate sizeEstimate = estimateMap.get(tokenRange);
+            // TODO: split range based on the partition size, not on the partition count
+            long partitionsCount = 0;
+            if (sizeEstimate != null) {
+                partitionsCount = sizeEstimate.getPartitionsCount();
             }
-            catch (Exception e) {
-                throw new IOException("Could not get input splits", e);
+            int subRangesCount = toIntExact((partitionsCount / splitSize) + 1);
+            List<TokenRange> subRanges = tokenRange.splitEvenly(subRangesCount);
```

Do you think we need to throw an exception when `.getMetadata().getTokenRanges()` returns empty set?